### PR TITLE
NCF model for adding feature

### PR DESCRIPTION
**Pull request by ForJadeForest**
## Description
This is the tensorflow ncf tutorial using spark dataframe to process data. I add more feature for the dataset.

### CODE REVIEW

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 1]
```diff
@@ -25,16 +27,35 @@ def read_data(data_dir):
...
+            StructField('item', IntegerType(), False),
+            StructField('title', StringType(), False),
+            StructField('category', StringType(), False)
+        ]
+    )
     # Need spark3 to support delimiter with more than one character.
-    df = spark.read.csv(
-        os.path.join(data_dir, "ratings.dat"), sep="::", schema=schema, header=False)
-    return df
+    df = spark.read.csv(os.path.join(data_dir, 'ratings.dat'),
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:47]: rename to df_rating?

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 2]
```diff
@@ -25,16 +27,35 @@ def read_data(data_dir):
...
     # Need spark3 to support delimiter with more than one character.
-    df = spark.read.csv(
-        os.path.join(data_dir, "ratings.dat"), sep="::", schema=schema, header=False)
-    return df
+    df_rating = spark.read.csv(os.path.join(data_dir, 'ratings.dat'),
+                        sep="::", schema=schema, header=False)
+    df_user = spark.read.csv(os.path.join(data_dir, 'users.dat'),
+                             sep="::", schema=schema_user, header=False)
+    df_item = spark.read.csv(os.path.join(data_dir, 'movies.dat'),
+                             sep="::", schema=schema_item, header=False).drop('title')
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:52]: can we remove drop just like ratings don't use rating column?

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 3]
```diff
@@ -58,14 +79,50 @@ def neg_sample(x):
...
     from bigdl.orca import init_orca_context, stop_orca_context
 
     sc = init_orca_context()
-    df = read_data('./ml-1m')
-    item_num = df.agg({'item': "max"}).collect()[0]["max(item)"]
-    df = generate_neg_sample(df, item_num)
-    train_data, test_data = df.randomSplit([0.8, 0.2], seed=100)
+
+    cat_feature = ['zipcode', 'gender', 'occupation', 'category']
+    num_feature = ['age']
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:122]: sparse_features and dense_features

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 4]
```diff
@@ -58,14 +79,50 @@ def neg_sample(x):
     return df
 
 
+def add_feature(df, df_user, df_item, cat_feature, num_feature):
+    df_feat = df.join(df_user, 'user', "inner")
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:83]: need to first transform cat features and then do the join to save cost

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 5]
```diff
@@ -58,14 +79,50 @@ def neg_sample(x):
...
+    df_feat = df_feat.join(df_item, 'item', "inner")
+    embedding_in_dim = []
+    for i in cat_feature:
+        indexer = StringIndexer(inputCol=i, outputCol=i + '_index').fit(df_feat)
+        df_feat = indexer.transform(df_feat)
+        df_feat = df_feat.drop(i).withColumnRenamed(i + '_index', i)
+        df_feat = df_feat.withColumn(i, df_feat[i].cast('int'))
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:90]: what's the type before cast?

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 6]
```diff
@@ -58,14 +79,50 @@ def neg_sample(x):
...
+        embedding_in_dim.append(df_feat.agg({i: "max"}).collect()[0][f"max({i})"])
+    for i in num_feature:
+        assembler = VectorAssembler(inputCols=[i], outputCol=i + '_vec')
+        scaler = MinMaxScaler(inputCol=i + '_vec', outputCol=i + '_scaler')
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:94]: scaler -> scaled

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 7]
```diff
@@ -58,14 +79,50 @@ def neg_sample(x):
...
+        pipeline = Pipeline(stages=[assembler, scaler])
+        scalerModel = pipeline.fit(df_feat)
+        df_feat = scalerModel.transform(df_feat)
+        df_feat = df_feat.drop(i, i + '_vec').withColumnRenamed(i + '_scaler', i)
+        df_feat.show()
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:99]: remove show?

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 8]
```diff
@@ -58,14 +79,50 @@ def neg_sample(x):
...
+    return df_feat, embedding_in_dim
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:100]: rename to sparse_feats_input_dims?

File: python/orca/tutorial/NCF/tf_model.py [hunk 9]
```diff
@@ -16,37 +16,57 @@
...
 
     with tf.name_scope("MLP"):
         user_embed_MLP = tf.keras.layers.Embedding(
-            user_num, embedding_size * 4, name='mlp_user')(user)
+            user_num, factor_num * 4, name='mlp_user')(user)
         item_embed_MLP = tf.keras.layers.Embedding(
-            item_num, embedding_size * 4, name='mlp_item')(item)
-        interaction = tf.concat([user_embed_MLP, item_embed_MLP], axis=-1)
+            item_num, factor_num * 4, name='mlp_item')(item)
+        interaction = tf.concat([user_embed_MLP, item_embed_MLP] + add_feature_layers, axis=-1)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:53]: shouldn't hard code 4, but use num_layers same as PyTorch

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 10]
```diff
@@ -58,14 +79,50 @@ def neg_sample(x):
...
+
+
+def data_process(data_dir, cat_feature, num_feature, neg_scale=4):
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:103]: rename to prepare_data, like xshards

File: python/orca/tutorial/NCF/tf_model.py [hunk 11]
```diff
@@ -16,37 +16,57 @@
 import tensorflow as tf
 
 
-def ncf_model(embedding_size, user_num, item_num, dropout, lr):
+def ncf_model(factor_num, user_num, item_num, dropout, lr,
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:19]: put factor_num after item_num

File: python/orca/tutorial/NCF/tf_model.py [hunk 12]
```diff
@@ -16,37 +16,57 @@
...
+              cat_features_in_dim, cat_features_out_dim, num_feature_dim):
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:20]: sparse_feats_input_dims and sparse_feats_embed_dims, like PyTorch

Comment (hkvision) [reply python/orca/tutorial/NCF/tf_model.py:20]: why num_feature_dim not num_dense_feats?

File: python/orca/tutorial/NCF/tf_model.py [hunk 13]
```diff
@@ -16,37 +16,57 @@
...
     user = tf.keras.layers.Input(dtype=tf.int32, shape=())
     item = tf.keras.layers.Input(dtype=tf.int32, shape=())
 
+    if not isinstance(cat_features_out_dim, list):
+        cat_features_out_dim = [cat_features_out_dim for _ in cat_features_in_dim]
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:25]: this is wrong, should be [cat_features_out_dim ] * len(cat_features_in_dim)?

File: python/orca/tutorial/NCF/tf_model.py [hunk 14]
```diff
@@ -16,37 +16,57 @@
...
+
+    cat_feature_input_layers = []
+    cat_feature_layers = []
+    for i, (in_dim, out_dim) in enumerate(zip(cat_features_in_dim, cat_features_out_dim)):
+        cat_feature_input_layers.append(tf.keras.layers.Input(shape=(), dtype=tf.int32))
+        cat_feature_layers.append(
+            tf.keras.layers.Embedding(in_dim + 1, out_dim)(cat_feature_input_layers[i]))
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:32]: do we need +1 in the model? should be handled before feeding the dim?

File: python/orca/tutorial/NCF/tf_model.py [hunk 15]
```diff
@@ -16,37 +16,57 @@
...
-            units=embedding_size * 2, activation='relu')(layer1_MLP)
+            units=factor_num * 2, activation='relu')(layer1_MLP)
         layer2_MLP = tf.keras.layers.Dropout(rate=dropout)(layer2_MLP)
         layer3_MLP = tf.keras.layers.Dense(
-            units=embedding_size, activation='relu')(layer2_MLP)
+            units=factor_num, activation='relu')(layer2_MLP)
         layer3_MLP = tf.keras.layers.Dropout(rate=dropout)(layer3_MLP)
 
     with tf.name_scope("concatenation"):
         concatenation = tf.concat([GMF, layer3_MLP], axis=-1)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:66]: are other features concated?

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 16]
```diff
@@ -58,14 +79,69 @@ def neg_sample(x):
...
+            df_item = num_scale(df_item, i)
+    df_feat = df.join(df_user, 'user', "inner")
+    df_feat = df_feat.join(df_item, 'item', "inner")
+    return df_feat, sparse_feats_input_dims
+
+
+def prepare_data(data_dir, neg_scale=4):
+    sparse_features = ['zipcode', 'gender', 'occupation', 'category']
+    dense_features = ['age']
+    total_features = ['user', 'item'] + sparse_features + dense_features
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:122]: rename to feature_cols?
And also put label_cols here?

File: python/orca/tutorial/NCF/tf_train_spark_dataframe.py [hunk 17]
```diff
@@ -28,33 +28,38 @@
...
+    sparse_feats_embed_dims=8
 )
 
 
 def model_creator(config):
-    model = ncf_model(embedding_size=config['embedding_size'],
-                      user_num=config['user_num'],
+    model = ncf_model(user_num=config['user_num'],
                       item_num=config['item_num'],
+                      num_layers=3,
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_train_spark_dataframe.py:51]: put num_layers in config?

File: python/orca/tutorial/NCF/tf_model.py [hunk 18]
```diff
@@ -16,37 +16,53 @@
...
-        user_embed_GMF = tf.keras.layers.Embedding(user_num, embedding_size, name='gmf_user')(user)
-        item_embed_GMF = tf.keras.layers.Embedding(item_num, embedding_size, name='gmf_item')(item)
+        user_embed_GMF = tf.keras.layers.Embedding(user_num, factor_num, name='gmf_user')(user)
+        item_embed_GMF = tf.keras.layers.Embedding(item_num, factor_num, name='gmf_item')(item)
         GMF = tf.keras.layers.Multiply()([user_embed_GMF, item_embed_GMF])
 
     with tf.name_scope("MLP"):
         user_embed_MLP = tf.keras.layers.Embedding(
-            user_num, embedding_size * 4, name='mlp_user')(user)
+            user_num, factor_num * 4, name='mlp_user')(user)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:34]: change to 4 to (2 ** (num_layers - 1))?

File: python/orca/tutorial/NCF/process_spark_dataframe.py [hunk 19]
```diff
@@ -58,14 +79,69 @@ def neg_sample(x):
...
+    df_rating, df_user, df_item = read_data(data_dir)
+
+    user_num = df_rating.agg({'user': "max"}).collect()[0]["max(user)"] + 1
+    item_num = df_rating.agg({'item': "max"}).collect()[0]["max(item)"] + 1
+
+    df_rating = generate_neg_sample(df_rating, item_num, neg_scale=neg_scale)
+    df_add_feature, sparse_feats_input_dims = \
+        add_feature(df_rating, df_user, df_item, sparse_features, dense_features)
+
+    train_df, val_df = df_add_feature.randomSplit([0.8, 0.2], seed=100)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/process_spark_dataframe.py:133]: rename df_add_feature to df?

File: python/orca/tutorial/NCF/tf_model.py [hunk 20]
```diff
@@ -16,37 +16,53 @@
...
+        num_feature_layers = []
+        for i, in_dim in enumerate(num_dense_feats):
+            num_feature_input_layers.append(tf.keras.layers.Input(shape=1))
+            num_feature_layers.append(num_feature_input_layers[i])
+
+        add_feature_input_layers = cat_feature_input_layers + num_feature_input_layers
+        add_feature_layers = cat_feature_layers + num_feature_layers
+
+        interaction = tf.concat([user_embed_MLP, item_embed_MLP] + add_feature_layers, axis=-1)
+        output_size = factor_num * (2 ** (num_layers - 1))
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:55]: add -> all?

File: python/orca/tutorial/NCF/tf_model.py [hunk 21]
```diff
@@ -16,37 +16,53 @@
...
+            interaction = tf.keras.layers.Dropout(rate=dropout)(layer_MLP)
+            output_size //= 2
 
     with tf.name_scope("concatenation"):
-        concatenation = tf.concat([GMF, layer3_MLP], axis=-1)
+        concatenation = tf.concat([GMF, interaction], axis=-1)
         outputs = tf.keras.layers.Dense(1, activation='sigmoid')(concatenation)
 
-    model = tf.keras.Model(inputs=[user, item], outputs=outputs)
+    model = tf.keras.Model(inputs=[user, item] + add_feature_input_layers, outputs=outputs)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:65]: wrong

File: python/orca/tutorial/NCF/tf_train_spark_dataframe.py [hunk 22]
```diff
@@ -28,33 +28,39 @@
 
 # Step 2: Read and process data using Spark DataFrame
 data_dir = './ml-1m'  # path to ml-1m
-df = read_data(data_dir)
-user_num = df.agg({'user': "max"}).collect()[0]["max(user)"] + 1
-item_num = df.agg({'item': "max"}).collect()[0]["max(item)"] + 1
-df = generate_neg_sample(df, item_num)
-train_df, val_df = df.randomSplit([0.8, 0.2], seed=100)
+
+train_df, val_df, sparse_feats_input_dims, user_num, item_num, total_features, label_cols = \
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_train_spark_dataframe.py:32]: change total_features to all_features

File: python/orca/tutorial/NCF/tf_model.py [hunk 23]
```diff
@@ -16,37 +16,53 @@
...
-            units=embedding_size, activation='relu')(layer2_MLP)
-        layer3_MLP = tf.keras.layers.Dropout(rate=dropout)(layer3_MLP)
+            item_num, factor_num * (2 ** (num_layers - 1)), name='mlp_item')(item)
+
+        cat_feature_input_layers = []
+        cat_feature_layers = []
+        for in_dim, out_dim in zip(sparse_feats_input_dims, sparse_feats_embed_dims):
+            cat_feature_input_layers.append(tf.keras.layers.Input(shape=(), dtype=tf.int32))
+            cat_feature_layers.append(
+                tf.keras.layers.Embedding(in_dim, out_dim)(cat_feature_input_layers[-1]))
```
Comment (hkvision) [inline python/orca/tutorial/NCF/tf_model.py:43]: change to
```
input_layer = tf.keras.layers.Input(shape=(), dtype=tf.int32)
cat_feature_input_layers.append(input_layer)
cat_feature_layers.append(tf.keras.layers.Embedding(in_dim, out_dim)(input_layer))
```
??
